proposals/0000-complete-set-signatures.rst

```diff
@@ -0,0 +1,483 @@
+
+ data Stream a = S a (Stream a)
+ instance IsList Stream where
+   viewCons (S x xs) = Just (x, xs)
+ instance IsInfinite Stream where
+
+ safeHead :: IsInfinite l => l a -> a
+ safeHead (x :< _) = x
+
+ {-# COMPLETE (:<) #-}
+
+Note that we declared the matching COMPLETE set in order not to be warned about
+``safeHead``.
+But in doing so, the following ``unsafeHead`` will not emit a pattern match
+warning anymore:
+
+::
+
+ unsafeHead :: [a] -> a
+ unsafeHead (x :< _) = x
+
+Urgh! We somehow want to say that the singleton COMPLETE set only applies
+to type constructors satisfying ``IsInfinite``. But the type signature
+syntax doesn't allow us to constrain on ``IsInfinite``! The only way out
+is to declare the COMPLETE signature for all concrete type constructors
+such as ``Stream``:
+
+::
+
+ {-# COMPLETE (:<) :: Stream #-}
+
+And here goes repeating that declaration for all type constructors again, for us
+as well as the users of our library. I'd much rather write
+
+::
+
+ {-# COMPLETE (:<), Empty :: IsList l => l a #-}
+ {-# COMPLETE (:<) :: IsInfinite l => l a #-}
+
+Once, inside the library. The constraint on the
+latter makes sure it doesn't apply to the match
+type in ``unsafeHead``, while it applies to any
+match type that satisfies ``IsInfinite`` that a
+user might write.
+
+That is the new feature that I propose. Note the
+specification of the full result type instead of
+just the name of a data type constructor: It's a
+breaking change with the current syntax, but one
+that is trivially flagged by a kinding error.
+
+Proposed Change Specification
+-----------------------------
+
+Syntax
+======
+
+Extend Syntax so that we are able to write the example from the previous section
+
+::
+
+ {-# COMPLETE (:<), Empty :: IsList l => l a #-} -- Context is allowed
+ {-# COMPLETE (:<) :: Stream a #-}               -- TyCon App is allowed
+ {-# COMPLETE (:<), Empty :: String #-}          -- Even type synonyms
+
+Since there is no formal grammar for COMPLETE pragmas, here's how I propose to
+change the happy grammar from
+
+::
+
+ sigdecl
+   : ...
+   | '{-# COMPLETE' con_list opt_tyconsig  '#-}'
+ opt_tyconsig
+   : {- -}
+   | '::' gtycon
+ gtycon
+   : ntgtycon
+   | '(' ')'
+   | '(#' '#)'
+
+to
+
+::
+
+ sigdecl
+   : ...
+   | '{-# COMPLETE' con_list opt_restysig  '#-}'
+ opt_restysig
+   : {- -}
+   | '::' ctype
+ atype
+   : ntgtycon
+   | '(' ')'
+   | '(#' '#)'
+   | ... many more ...
+
+Where a ``ctype`` is a "for-all type", according to GHC's happy parser. It's the
+same grammatical sort that constitutes the RHS of a ``::`` in a type signature
+and since it constrains the result type of the relevant ConLikes, we call
+it a "result type signature", often referring just to the type it carries.
+
+Note that this change accepts strictly more syntax, because
+
+- ``atype`` accepts a superset of ``gtycon``
+- ``atype`` can be derived from ``ctype`` (via ``type``, ``btype``, ``infixtype``, ``ftype``)
+
+Typing
+======
+
```

Comment:
Is this ok?
```
{-# COMPLETE Just :: Int #-} 
```
Presumably not.  Somehow the result type of the constructor should be compatible with `sig_ty`.

Reply:
Yes. I would say that the result type of the pattern must be more polymorphic than the `sig_ty`. (Here, as everywhere, "more polymorphic than" is a reflexive relation, because writing "more polymorphic than or equivalent to" is a mouthful.)